### PR TITLE
MBS-11769 add umbrella tasks

### PR DESCRIPTION
**Pull request by dsvoronin**
samples instrumentation tasks was in separate teamcity build, now it's a part of :checkAll / :build tasks, so executed in CI single build

move documentation lint before gradle build

todo:
 - remove all `-p subprojects`
 - compileAll for included build is now stub task

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -15,10 +16,55 @@ tasks.register<Exec>("installGitHooks") {
...
+    dependsOn(
+        gradle.includedBuilds
+            .filter { !it.name.startsWith("build-logic") }
+            .map { it.task(":compileAll") }
+    )
+}
+
+val assembleAll = tasks.register("assembleAll") {
+    group = taskGroup
+    description = "Compile all code of all components"
```
Comment (eugene-krivobokov) [inline build.gradle.kts]:
```suggestion
    description = "Assemble all components"
```

File: build.gradle.kts [hunk 2]
```diff
@@ -15,10 +16,55 @@ tasks.register<Exec>("installGitHooks") {
...
+        gradle.includedBuilds
+            .filter { !it.name.startsWith("build-logic") }
+            .map { it.task(":assembleAll") }
+    )
+}
+
+tasks.register("build") {
+    group = taskGroup
+    description = "Run all tests, build (without publishing)"
+    dependsOn(assembleAll, compileAll, checkAll)
```
Comment (eugene-krivobokov) [inline build.gradle.kts]:
```suggestion
    dependsOn(assembleAll, checkAll)
```
Do we have cases when assemble doesn't include compile?

File: build.gradle.kts [hunk 3]
```diff
@@ -15,10 +16,55 @@ tasks.register<Exec>("installGitHooks") {
...
+    dependsOn(
+        gradle.includedBuilds
+            .filter { !it.name.startsWith("build-logic") }
+            .map { it.task(":assembleAll") }
+    )
+}
+
+tasks.register("build") {
+    group = taskGroup
+    description = "Run all tests, build (without publishing)"
```
Comment (eugene-krivobokov) [inline build.gradle.kts]:
```suggestion
    description = "Build and run all tests (without publishing)"
```
To reflect real execution order

File: build.gradle.kts [hunk 4]
```diff
@@ -15,10 +16,55 @@ tasks.register<Exec>("installGitHooks") {
...
+
+// Register lifecycle tasks in this umbrella build.
+// A user/CI usually only needs these.
+val checkAll = tasks.register("checkAll") {
+    group = taskGroup
+    description = "Run all tests and create code coverage report"
+
+    dependsOn(
+        gradle.includedBuilds
+            .filter { !it.name.startsWith("build-logic") }
```
Comment (eugene-krivobokov) [inline build.gradle.kts]: Can we hide this duplicated knowledge about "build-logic"?
Maybe reusable function or make fake empty tasks inside build-logic.

File: build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+val taskGroup = "Avito android build"
```
Comment (eugene-krivobokov) [inline build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts]:
```suggestion
val taskGroup = "Avito Android build"
```
Just to be consistent as in [official guidelines](https://developer.android.com/distribute/marketing-tools/brand-guidelines#android_in_text)

File: build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+tasks.named<TaskReportTask>("tasks") {
+    displayGroup = taskGroup
```
Comment (eugene-krivobokov) [inline build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts]: How could I see tasks from others groups?

File: build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+tasks.register("assembleAll") {
+    group = taskGroup
+    description = "Compile all code of the '${project.name}' component"
```
Comment (eugene-krivobokov) [inline build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts]:
```suggestion
    description = "Compile all code of the '${project.path}' component"
```
Do we need to worry about probable duplicates here?

File: build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    group = taskGroup
+    description = "Compile all code and run all tests of the '${project.name}' component"
+    dependsOn(
+        subprojects
+            .filter { !it.isJustADir() }
+            .map { "${it.path}:check" }
+    )
+}
+
+fun Project.isJustADir(): Boolean {
```
Comment (eugene-krivobokov) [inline build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts]:
```suggestion
fun Project.isPlainDir(): Boolean
```
```suggestion
fun Project.isModule(): Boolean
```
Just as options

File: build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts (hunk 7 above)
Comment (dsvoronin) [reply build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts]: In description probably not, but i will use path anyway

File: build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts (hunk 6 above)
Comment (dsvoronin) [reply build-logic/gradle/src/main/kotlin/convention.lifecycle.gradle.kts]: I don't get why it's here 😆 
Just copy pasted from an example

removed

File: build.gradle.kts (hunk 2 above)
Comment (dsvoronin) [reply build.gradle.kts]: compileAll also includes compiling unit and android tests, which should be a part of checkAll

File: build.gradle.kts (hunk 4 above)
Comment (dsvoronin) [reply build.gradle.kts]: i made tasks in build-logic projects

**Pull request merged**